lib/RestructureCFG/InlineDivergentScopesPass.cpp

```diff
@@ -334,24 +376,123 @@ static void performIDS(DivergenceDescriptor &DivergenceDescriptor,
   simplifyTerminator(Tail, PlaceHolderTarget);
 
   // We add a `scope_closer` edge between the divergent exit node and
-  // the `Tail` node
-  ScopeGraphBuilder SGBuilder(F);
-  SGBuilder.addScopeCloser(Exit, Tail);
+  // the `Tail` node for all the exits
+  for (auto Divergence : MultipleDivergences) {
+    SGBuilder.addScopeCloser(Divergence.Exit, Tail);
+  }
+}
+
+/// Helper function which returns if a `DivergenceDescriptor` is compatible with
+/// all the ones already present in a `Collection`. Being compatible means that
+/// they insist on the same conditional, have non-overlapping `Exit` blocks, and
+/// do not have any overlapping `DivergentSuccessors`
+static bool
+canAddDescriptor(const SmallVector<DivergenceDescriptor> &Collection,
+                 const DivergenceDescriptor &NewDescriptor) {
+  // Skip checking if collection is empty
+  if (Collection.empty())
+    return true;
+
+  // Check that Conditional matches
+  if (Collection[0].Conditional != NewDescriptor.Conditional)
+    return false;
+
+  // Check for non-overlapping Exit blocks
+  for (const auto &Existing : Collection) {
+    if (Existing.Exit == NewDescriptor.Exit)
+      return false;
+
+    // Check for non-overlapping DivergentSuccessors
+    for (BasicBlock *ExistingSucc : Existing.DivergentSuccessors) {
+      for (BasicBlock *NewSucc : NewDescriptor.DivergentSuccessors) {
+        if (ExistingSucc == NewSucc)
+          return false;
+      }
+    }
+  }
+
+  return true;
+}
+
+/// Helper function which attempts to perform multiple IDS transformations
+static bool
+tryMultipleIDS(ScopeGraphBuilder &SGBuilder,
+               SmallVector<DivergenceDescriptor> &DivergenceDescriptors,
+               BasicBlock *PlaceHolderTarget) {
+
+  // We attempt to perform multiple IDS transformations all at once
+  // We start by collecting all the `Conditional`s involved in at least one
+  // divergence
+  SmallSet<BasicBlock *, 4> Conditionals;
+  for (auto DivergenceDescriptor : DivergenceDescriptors) {
+    Conditionals.insert(DivergenceDescriptor.Conditional);
+  }
+
+  // For each conditional, we try to collect all the multiple divergences that
+  // involves such `Conditional`
+  for (BasicBlock *Conditional : Conditionals) {
```

Comment:
The key of `Conditionals` is a pointer, so this iteration is non-deterministic. Make `Conditionals` a `SetVector` please.

Reply:
Oh good catch, changed this here https://github.com/revng/revng/blob/3a9379ef740d1d2b9f7bcad02cc962719251e02a/lib/RestructureCFG/InlineDivergentScopesPass.cpp#L333C3-L333C20

Reply:
Also you always do this once, for the first `Conditional`. The loop is pointless, just take `Conditionals.begin()`.